### PR TITLE
FAI 795v2: Fixed Invalid-Nonbest CF results

### PR DESCRIPTION
**Pull request by RobGeada**
https://issues.redhat.com/browse/FAI-795

Optaplanner clones a single CounterfactualSolution object for each candidate input to the model. Previously, the `entities` attribute of the CFSolutions were not properly marked as Optaplanner `PlanningEntities`, which meant OP was not cloning the correct entities when recalling the best solution, instead seemingly picking some arbitrary entities from the stack. To make this change work, all FixedEntities were removed, as the `@PlanningPin` annotation allows for the marking of normal Entities as "pinned", which lets both fixed and variable entities be correctly saved by OP for later restoratation.

Additionally, this change breaks the diversifier tests: the found CFs are now too sparse and close to the decision boundary for the current diversifier implementation to work, odds are it will need multiple CFs to inter-permute before it will work again. For now, the tests are disabled.

### CODE REVIEW

**Review by tteofili — APPROVED** [review]
LGTM

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifier.java [hunk 1]
```diff
@@ -167,6 +167,7 @@ public Dataset generateAlternatives() throws ExecutionException, InterruptedExce
         final Set<List<Feature>> uniqueFeatures = new HashSet<>();
         final List<Prediction> predictions = new ArrayList<>();
         final double[] probabilities = generateProbabilities();
+        int totalTries = 0;
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifier.java]: Leftover? It is just updated but never checked

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractEntity.java [hunk 2]
```diff
@@ -17,7 +17,11 @@
 
 import java.util.Objects;
 
+import javax.enterprise.inject.Default;
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractEntity.java]: Are you sure about this? This annotation is used mainly for dependency injection, is OptaPlanner using this annotation for something different?

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/SolverConfigBuilder.java [hunk 3]
```diff
@@ -68,6 +69,21 @@ public SolverConfig build() {
...
+                    //                    FixedCurrencyEntity.class,
+                    //                    FixedDoubleEntity.class,
+                    //                    FixedDurationEntity.class,
+                    //                    FixedIntegerEntity.class,
+                    //                    FixedLongEntity.class,
+                    //                    FixedObjectEntity.class,
+                    //                    FixedTextEntity.class,
+                    //                    FixedTimeEntity.class,
+                    //                    FixedURIEntity.class,
+                    //                    FixedVectorEntity.class,
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/SolverConfigBuilder.java]: To be removed? If yes please remove also the actual classes :)

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractCategoricalEntity.java [hunk 4]
```diff
@@ -52,9 +57,18 @@ public double similarity() {
...
 
-    public abstract Set<T> getValueRange();
+    @Override
+    public abstract CountableValueRange<T> getValueRange();
+
+    public Set<T> recoverSet() {
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(
+                getValueRange().createOriginalIterator(), Spliterator.ORDERED), false)
+                .collect(Collectors.toSet());
+    }
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractCategoricalEntity.java:73]: Deadcode? Or maybe `@Override` also here?

File: explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java [hunk 5]
```diff
@@ -47,33 +51,41 @@ class CounterfactualDiversifierTest {
...
+        System.out.println(MatrixUtilsExtensions.vectorFromPredictionInput(
+                new PredictionInput(
+                        result.getEntities().stream().map(CounterfactualEntity::asFeature).collect(Collectors.toList()))));
+        System.out.println(result.isValid());
         for (CounterfactualEntity entity : result.getEntities()) {
             totalSum += entity.asFeature().getValue().asNumber();
             logger.debug("Entity: {}", entity);
         }
-
+        System.out.println(result.getOutput().get(0).getOutputs());
```
Comment (danielezonca) [inline explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java]: Leftover

File: explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java [hunk 6]
```diff
@@ -92,6 +104,9 @@ void simpleDiverse(int seed) throws ExecutionException, InterruptedException, Ti
         }
     }
 
+    // these tests fail to find diverse counterfactuals; the original CF is too sparse + close to decision
+    // boundary for the diversifier to work
+    @Disabled
```
Comment (danielezonca) [inline explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java]: Please add ticket number
```suggestion
    @Disabled("TICKET")
```

File: explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java [hunk 7]
```diff
@@ -47,33 +51,41 @@ class CounterfactualDiversifierTest {
     private static final Logger logger =
             LoggerFactory.getLogger(CounterfactualExplainerTest.class);
 
+    // these tests fail to find diverse counterfactuals; the original CF is too sparse + close to decision
+    // boundary for the diversifier to work
+    @Disabled
```
Comment (danielezonca) [inline explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java]: Same here

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractCategoricalEntity.java (hunk 4 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractCategoricalEntity.java:73]: This is used in the unit tests to ensure categorical entities are being correctly instantiated. It's only applicable to this class its subclasses, the parent and sibling classes cannot use this function so Override would not make sense here.

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/SolverConfigBuilder.java (hunk 3 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/SolverConfigBuilder.java]: done

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractEntity.java (hunk 2 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractEntity.java]: fixed

File: explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java (hunk 5 above)
Comment (RobGeada) [reply explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java]: fixed

File: explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java (hunk 6 above)
Comment (RobGeada) [reply explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java]: done

File: explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java (hunk 7 above)
Comment (RobGeada) [reply explainability-core/src/test/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifierTest.java]: done

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifier.java (hunk 1 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifier.java]: is indeed leftover, removed.

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractCategoricalEntity.java (hunk 4 above)
Comment (danielezonca) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/entities/AbstractCategoricalEntity.java:73]: Ok thanks for the clarification, I was not able to find any usage so I was guessing if it was just dead code :)

**Review by danielezonca — APPROVED** [review]
Thanks for the changes, LGTM 👍

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualSolution.java [hunk 8]
```diff
@@ -36,11 +36,10 @@
  */
 @PlanningSolution
 public class CounterfactualSolution {
-
+    @PlanningEntityCollectionProperty
     private List<CounterfactualEntity> entities;
     private List<Feature> originalFeatures;
 
-    @PlanningEntityCollectionProperty
     public List<CounterfactualEntity> getVaryingEntities() {
```
Comment (ruivieira) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualSolution.java]: Given that the entities are not accessed using `getVaryingEntities` anymore (and this doesn't seem to be used) could it be removed, perhaps?

**Review by ruivieira — APPROVED** [review]
LGTM, thanks @RobGeada.

Regarding the diversifier, as we discussed, this is not a problem, since it is meant to be used (in the future) with multiple CFs as the inputs. At the moment, multiple CFs is not yet implemented, so it's fine to disable the tests, AFAIC.

File: explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualSolution.java (hunk 8 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualSolution.java]: Yep, will do

**Pull request merged**